### PR TITLE
Scale projectile damage based on weapon quality

### LINKED ISSUE #1867: [Discussion]: Legendary and Masterwork guns
Since guns in CE don't benefit from quality as much as in vanilla for some guns even less than others, especially damage-wise, it might be a good idea to introduce a slight damage boost (5% and 10% for legendary and masterwork quality respectively) to high quality guns to simulate a better muzzle velocity resulting in more damage dealt by the gun. 
The current accuracy boost does not matter for precision guns which are already almost capped accuracy-wise and quests can have high quality guns as rewards which are sub-par in comparison to other rewards.

This'd make spending inspirations on guns a bit more viable in comparison to armor pieces which scale greatly with quality.

### PR DESCRIPTION
**Pull request by mszabo-wikia**
## Additions

Apply a multiplier to bullet damage based on the quality of the weapon that fired it, just like vanilla does. This uses the RangedWeapon_DamageMultiplier vanilla stat.

For this, move the DamageAmount into the base ProjectileCE class from
BulletCE so that all ProjectileCE subclasses can use it to obtain a
correct damage amount scaled by launcher quality.

Awful guns deal x0.9 damage, Masterwork ones deal x1.05 and Legendary ones deal x1.1 damage now.

## References

- Closes #1867

## Testing

Check tests you have performed:
- [x] Compiles without warnings
- [x] Game runs without errors
- [x] Playtested a colony (working as intended)

### CODE REVIEW

**Comment by SamaelGray** [comment]
Works as intended in-game, will approve after the last 1.3 release.

**Comment by T-rexsuper1** [comment]
holy crap is this a dream

**Review by SamaelGray — APPROVED** [review]

**Review by SamaelGray — DISMISSED** [review]

**Review by SamaelGray — DISMISSED** [review]

File: Source/CombatExtended/CombatExtended/Lasers/LaserBeamCE.cs [hunk 1]
```diff
@@ -11,11 +11,11 @@ namespace CombatExtended.Lasers
     public class LaserBeamCE : BulletCE
     {
 
-	public override float DamageAmount
+        public override int DamageAmount
```
Comment (perkinslr) [inline Source/CombatExtended/CombatExtended/Lasers/LaserBeamCE.cs:14]: Why is this being converted *back* to an int?  We'll likely be moving all bullet damage to floats eventually, no reason to add more work later for lasers.

File: Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs [hunk 2]
```diff
@@ -93,6 +93,34 @@ public Thing intendedTargetThing
...
 
+        /// <summary>
+        /// Backing field for <see cref="DamageAmount"/>.
+        /// </summary>
+        private int? damageAmount;
+
+        /// <summary>
+        /// Return the damage dealt by this projectile scaled by the quality multiplier of its launcher.
+        /// </summary>
+        public virtual int DamageAmount
```
Comment (perkinslr) [inline Source/CombatExtended/CombatExtended/Projectiles/ProjectileCE.cs:104]: Should return a float.  We use floats internally for damage calculations.  It's a limitation of Core that base weapon damage are integers, but this is an ideal place to make them floats, and will make converting the base damage to floats in the future easier.

**Review by perkinslr — CHANGES_REQUESTED** [review]

**Review by SamaelGray — APPROVED** [review]

**Review by Thorbane — APPROVED** [review]

**Pull request merged**